L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorSuperTriggerCellImpl.cc

```diff
@@ -23,21 +24,87 @@ HGCalConcentratorSuperTriggerCellImpl(const edm::ParameterSet& conf)
 
 int
 HGCalConcentratorSuperTriggerCellImpl::getSuperTriggerCellId(int detid) const {
-  // FIXME: won't work in the V9 geometry
-  HGCalDetId TC_id(detid);
-  if(TC_id.subdetId()==HGCHEB) {
-    return TC_id.cell(); //scintillator
-  } else {
 
-    int TC_wafer = TC_id.wafer();
-    int TC_12th = ( TC_id.cell() & kSplit12_ );
-    int TC_3rd = ( TC_id.cell() & kSplit3_ );
 
-    int thickness = triggerTools_.thicknessIndex(detid,true);
-    int TC_split = TC_12th;
-    if (stcSize_.at(thickness) == kSTCsize16_) TC_split = TC_3rd;
+  DetId TC_id( detid );
+  if ( TC_id.det() == DetId::Forward ){//V8
 
-    return TC_wafer<<kWafer_offset_ | TC_split;
+    HGCalDetId TC_idV8(detid);
+
+    if( triggerTools_.isScintillator(detid) ){
+      return TC_idV8.cell(); //scintillator
+    }
+    else{
+      int TC_wafer = TC_idV8.wafer();
+      int TC_12th = ( TC_idV8.cell() & kSplit12_ );
+      int TC_3rd = ( TC_idV8.cell() & kSplit3_ );
+      
+      int thickness = triggerTools_.thicknessIndex(detid,true);
+      int TC_split = TC_12th;
+      if (stcSize_.at(thickness) == kSTCsize16_) TC_split = TC_3rd;
+      
+      return TC_wafer<<kWafer_offset_ | TC_split;
+    }
+
+  }
+
+  else if ( TC_id.det() == DetId::HGCalTrigger ){//V9
+    
+    HGCalTriggerDetId TC_idV9(detid);
+
+    if( triggerTools_.isScintillator(detid) ){
+      return TC_idV9.triggerCellU() << kWafer_offset_ | TC_idV9.triggerCellV(); //scintillator
+    }
+    else {
+      int TC_wafer = TC_idV9.waferU() << kWafer_offset_ | TC_idV9.waferV() ;
+      int thickness = triggerTools_.thicknessIndex(detid);
+
+
+      int TC_cell = TC_idV9.triggerCellU() << kThree_ | TC_idV9.triggerCellV();
+
+      int TC_12th = 0;
+      int TC_3rd = 1;
+
+      if ( TC_idV9.triggerCellU() > kThree_ || TC_idV9.triggerCellV() > kThree_ ){
```

Comment:
Also, all the three diamonds should be equivalent, except for the rotation. So the same TC->STC association could be used, with a (u,v) rotation before.
For instance in the picture below in the top-right diamond (starting at 4,4), the grouping can simply be done with something like `U=(u-4)/2` and `V=(v-4)/2`.

![image](https://user-images.githubusercontent.com/5569798/54610303-08185a00-4a55-11e9-9ec1-fc4719908774.png)

Then the other diamonds can be treated in the exact same way with just a redefinition of u and v (a rotation + a shift).

We can discuss it over Skype if it is not clear or if I'm missing something.

Reply:
I think this is essentially what I have now, I use the same value of "kSplit12_v9" for each diamond, but first modify the u and v values. I will look into using the ROC number.

Reply:
Sorry, but it is not clear at all. Could you make it more readable?
First define your new `u'` and `v'` depending on the case, then use these these `u'` and `v'` for all the different cases with a single statement, to avoid code duplication.

Reply:
Sure, I will.
I see that the ROC number is an integer between 1 and 6, but those numbers do not uniquely correspond to certain diamonds. Do you know what the correspondence is?

Reply:
It would actually be better to use [getROCNumber(int triggerCellU, int triggerCellV, int type)](https://github.com/cms-sw/cmssw/blob/master/DataFormats/ForwardDetId/interface/HGCSiliconDetIdToROC.h#L21), with `type=1`; it should give ROCs corresponding to coarse wafers (= diamonds).

Reply:
This does give integers in the range 1-3 but they do not uniquely correspond to the division I have made using the U and V values (which I think is correct).

Reply:
Actually with the numbering scheme in Sunanda's presentation, your splitting doesn't seem correct (it does not correspond to the expected diamonds or the ROCs).

Reply:
There was a problem with my numbering system, fixed.